File: examples/androidJavaSDKExample/app/src/main/java/com/llamamobile/sdkexample/EmbeddingsFragment.java
package com.llamamobile.sdkexample;

import android.os.Bundle;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.TextView;
import android.widget.EditText;
import android.widget.Button;
import android.widget.Toast;
import android.widget.ProgressBar;

import androidx.fragment.app.Fragment;

import com.llamamobile.sdk.LlamaMobileSdk;

import java.util.List;

public class EmbeddingsFragment extends Fragment {

    private EditText embedEditText;
    private Button generateEmbeddingsButton;
    private TextView embeddingsTextView;
    private ProgressBar progressBar;

    private boolean isProcessing = false;

    public EmbeddingsFragment() {
        // Required empty public constructor
    }

    @Override
    public View onCreateView(LayoutInflater inflater, ViewGroup container,
                             Bundle savedInstanceState) {
        View rootView = inflater.inflate(R.layout.fragment_embeddings, container, false);

        embedEditText = rootView.findViewById(R.id.embedEditText);
        generateEmbeddingsButton = rootView.findViewById(R.id.generateEmbeddingsButton);
        embeddingsTextView = rootView.findViewById(R.id.embeddingsTextView);
        progressBar = rootView.findViewById(R.id.progressBar);

        generateEmbeddingsButton.setOnClickListener(v -> handleGenerateEmbeddings());

        updateProcessingUI();

        return rootView;
    }

    private void handleGenerateEmbeddings() {
        String text = embedEditText.getText().toString().trim();
        if (text.isEmpty() || isProcessing) {
            return;
        }

        isProcessing = true;
        updateProcessingUI();

        MainActivity mainActivity = (MainActivity) getActivity();
        if (mainActivity != null) {
            AppState appState = mainActivity.getAppState();
            LlamaMobileSdk llamaMobileSdk = appState.getLlamaMobileSdk();
            llamaMobileSdk.generateEmbeddings(text, new LlamaMobileSdk.ResultCallback<List<Float>>() {
                @Override
                public void onSuccess(List<Float> result) {
                    getActivity().runOnUiThread(() -> {
                        isProcessing = false;
                        embeddingsTextView.setText(formatEmbeddings(result));
                        updateProcessingUI();
                    });
                }

                @Override
                public void onError(Exception e) {
                    getActivity().runOnUiThread(() -> {
                        isProcessing = false;
                        Toast.makeText(getContext(), "Embeddings error: " + e.getMessage(), Toast.LENGTH_SHORT).show();
                        updateProcessingUI();
                    });
                }
            });
        }
    }

    private String formatEmbeddings(List<Float> embeddings) {
        StringBuilder sb = new StringBuilder();
        sb.append("Embeddings (Dimension: " + embeddings.size() + "):\n");
        int maxDisplay = Math.min(10, embeddings.size());
        for (int i = 0; i < maxDisplay; i++) {
            sb.append(String.format("%.4f", embeddings.get(i)));
            if (i < maxDisplay - 1) {
                sb.append(", ");
            }
        }
        if (embeddings.size() > maxDisplay) {
            sb.append("...");
        }
        return sb.toString();
    }

    private void updateProcessingUI() {
        if (getView() == null) return;

        progressBar.setVisibility(isProcessing ? View.VISIBLE : View.GONE);
        generateEmbeddingsButton.setEnabled(!isProcessing);
    }
}

File: examples/androidJavaSDKExample/app/src/main/java/com/llamamobile/sdkexample/EmbeddingsFragment.kt
package com.llamamobile.sdkexample

import android.os.Bundle
import android.view.LayoutInflater
import android.view.View
import android.util.Log
import android.view.ViewGroup
import android.content.Context
import android.widget.Toast
import androidx.fragment.app.Fragment
import com.llamamobile.LlamaMobile
import com.llamamobile.sdkexample.databinding.FragmentEmbeddingsBinding

class EmbeddingsFragment : Fragment() {

    private var _binding: FragmentEmbeddingsBinding? = null
    private val binding get() = _binding!!
    private var appState: AppState? = null
    private var isProcessing = false

    override fun onAttach(context: Context) {
        super.onAttach(context)
        // Access appState when the fragment is properly attached to the activity
        appState = (context as? MainActivity)?.appState
    }

    override fun onCreateView(
        inflater: LayoutInflater,
        container: ViewGroup?,
        savedInstanceState: Bundle?
    ): View {
        _binding = FragmentEmbeddingsBinding.inflate(inflater, container, false)
        return _binding!!.root
    }

    override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
        super.onViewCreated(view, savedInstanceState)

        try {
            // Set default text
            binding.embedEditText.setText("This is a test sentence")

            // Set up generate embeddings button click listener
            binding.generateEmbeddingsButton.setOnClickListener {
                handleGenerateEmbeddings()
            }

            // Update UI based on model loading status
            updateModelLoadedUI()
        } catch (e: Exception) {
            // Log but don't crash if there's an issue with embeddings fragment initialization
            e.printStackTrace()
        }
    }

    override fun onResume() {
        super.onResume()
        // Check model status when fragment resumes
        updateModelLoadedUI()
    }
    
    override fun onDestroyView() {
        super.onDestroyView()
        _binding = null
    }

    private fun handleGenerateEmbeddings() {
        val text = binding.embedEditText.text.toString().trim()
        if (text.isEmpty() || isProcessing) {
            return
        }

        val currentAppState = appState ?: run {
            activity?.let { Toast.makeText(it, "App state not initialized", Toast.LENGTH_SHORT).show()}
            return
        }

        if (!currentAppState.isModelLoaded) {
            activity?.let { Toast.makeText(it, "Please load a model first", Toast.LENGTH_SHORT).show()}
            return
        }

        if (!currentAppState.enableEmbedding) {
            activity?.let { Toast.makeText(it, "Embedding is not enabled. Please enable it in Settings and reload the model.", Toast.LENGTH_SHORT).show()}
            return
        }

        isProcessing = true
        updateProcessingUI()

        // Generate embeddings in a background thread
        Thread {
            try {
                // Debug embedding generation
                Log.d("EmbeddingFragment", "- Context handle: ${currentAppState.contextHandle}")
                Log.d("EmbeddingFragment", "- Text: $text")
                
                // Call the new LlamaMobile generateEmbeddings method
                val embeddingsArray = LlamaMobile.generateEmbeddings(currentAppState.contextHandle, text)
                
                Log.d("EmbeddingFragment", "- Embeddings result: ${embeddingsArray?.size ?: "null"}")
                
                activity?.runOnUiThread {
                    isProcessing = false
                    updateProcessingUI()
                    
                    if (embeddingsArray != null) {
                        val embeddingsList = embeddingsArray.toList()
                        binding.embeddingsTextView.text = formatEmbeddings(embeddingsList)
                        
                        // Show embeddings result section
                        binding.embeddingsSection.visibility = View.VISIBLE
                    } else {
                        activity?.let { Toast.makeText(it, "Embeddings generation failed: null result", Toast.LENGTH_SHORT).show()}
                        
                        // Hide embeddings result section on error
                        binding.embeddingsSection.visibility = View.GONE
                    }
                }
            } catch (e: Exception) {
                activity?.runOnUiThread {
                    isProcessing = false
                    updateProcessingUI()
                    activity?.let { Toast.makeText(it, "Embeddings error: ${e.localizedMessage}", Toast.LENGTH_SHORT).show()}
                    
                    // Hide embeddings result section on error
                    binding.embeddingsSection.visibility = View.GONE
                }
            }
        }.start()
    }

    private fun formatEmbeddings(embeddings: List<Float>): String {
        if (embeddings.isEmpty()) {
            return "No embeddings generated"
        }

        val truncated = embeddings.take(20)
        val formatted = truncated.joinToString(", ") { "%.6f".format(it) }
        return if (embeddings.size > 20) {
            "$formatted, ... (and ${embeddings.size - 20} more values)"
        } else {
            formatted
        }
    }

    private fun updateModelLoadedUI() {
        val modelLoaded = appState?.isModelLoaded ?: false
        val embeddingEnabled = appState?.enableEmbedding ?: false
        binding.generateEmbeddingsButton.isEnabled = modelLoaded && !isProcessing && embeddingEnabled

        if (!modelLoaded) {
            binding.statusTextView.text = "Model not loaded"
            binding.statusTextView.setTextColor(android.graphics.Color.RED)
        } else if (!embeddingEnabled) {
            binding.statusTextView.text = "Embedding disabled. Enable in Settings and reload model."
            binding.statusTextView.setTextColor(android.graphics.Color.RED)
        } else {
            binding.statusTextView.text = "Model loaded and embedding enabled"
            binding.statusTextView.setTextColor(android.graphics.Color.GREEN)
        }
    }

    private fun updateProcessingUI() {
        binding.generateEmbeddingsButton.isEnabled = (appState?.isModelLoaded ?: false) && !isProcessing && (appState?.enableEmbedding ?: false)
        binding.progressBar.visibility = if (isProcessing) View.VISIBLE else View.GONE
    }
}
